### PR TITLE
Distinguish between public and private notifications.

### PR DESCRIPTION
**Pull request by dennisguse**
Add a public notification (which does not show the current title and artist).
If the lock screen is configured to _hide sensitive information_, this notification will be shown instead.

This allows to use interact with Odyssey from the lock screen while preventing others to see the currently playing song.

### CODE REVIEW

**Comment by gnome17** [comment]
Hi,
interesting proposal. We think it could be argued that these information should be private but we don't share this opinion. So maybe we could add an option for this but not as a default.

**Comment by dennisguse** [comment]
Sounds reasonable.
I will update the change request in the upcoming days. 

On January 31, 2018 5:25:29 PM UTC, "Frederik Lütkes" <notifications@github.com> wrote:
>Hi,
>interesting proposal. We think it could be argued that these
>information should be private but we don't share this opinion. So maybe
>we could add an option for this but not as a default.
>
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/gateship-one/odyssey/pull/120#issuecomment-362006692

-- 
Dennis Guse

**Comment by dennisguse** [comment]
IMPORTANT: the preference patch has a serious limitation!

It requires restarting the PlaybackService if the "private notification"-preference is changed.
I could figure out, how this is done correctly...
Any suggestions?

**Comment by gnome17** [comment]
Hi,
the problem are the SharedPreferences (see https://developer.android.com/reference/android/content/SharedPreferences.html) :
> Note: This class does not support use across multiple processes.

So you have to do something similar to the method hideArtwork in the PlaybackService. This will require more changes including extending the aidl file.

**Comment by dennisguse** [comment]
Works now correctly :)

I amended the changes to last commit.

Just one more thing: if a notification exists, I updated this notification using `mNotification.visibility(...)`.
I tried rebuilding it using the NotificationBuilder, but this removed the controls when going private.
I don't know why, but I think it would be the better suited option.

I included both version (NotificationBuilder commented out) - It is marked with a comment.
Just search for "ATTENTION".

What do you say?

**Comment by dennisguse** [comment]
From my side is code ready for inclusion.

I revisited again the issue on how to update the notification visibility (a: create a new one notification or b: update the visibility).
I updated the current notification as the NotificationBuilder cannot be reused later.
This will remove the playback controls, which would be quite annoying.

Whenever you have time.

File: app/src/main/java/org/gateshipone/odyssey/fragments/SettingsFragment.java [hunk 1]
```diff
@@ -180,6 +180,14 @@ public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, Strin
             getActivity().finish();
             startActivity(intent);
         }
+
+        if (key.equals(getString(R.string.pref_notification_private_key))) {
+            try {
+                boolean notificationPrivate = sharedPreferences.getBoolean(key, getResources().getBoolean(R.bool.pref_notification_private_default));
+                mServiceConnection.getPBS().notificationPrivateChanged(notificationPrivate);
```
Comment (gnome17) [inline app/src/main/java/org/gateshipone/odyssey/fragments/SettingsFragment.java]: We had to change the connection handling and removed the mServiceConnection member. Please use the getPlaybackService method from the GenericActivity instead.

File: app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java [hunk 2]
```diff
@@ -400,6 +402,11 @@ public void hideArtwork(boolean enable) {
         updateStatus();
     }
 
+    public void notificationPrivate(boolean enable) {
+        mNotificationManager.notificationPrivate(enable);
+        //updateStatus(); //This creates a null pointer exception!
```
Comment (gnome17) [inline app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java]: Is there a reason why this wasn't removed or the npe resolved?

File: app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/OdysseyNotificationManager.java [hunk 3]
```diff
@@ -143,6 +146,11 @@ public synchronized void updateNotification(TrackModel track, PlaybackService.PL
             notificationStyle.setShowActionsInCompactView(1, 2);
             notificationStyle.setMediaSession(mediaSessionToken);
             mNotificationBuilder.setStyle(notificationStyle);
+            mNotificationBuilder.setContentTitle(mContext.getResources().getString(R.string.notification_sensitive_content_replacement));
+
+            //Build the public notification
+            Notification notificationPublic = mNotificationBuilder.build();
```
Comment (gnome17) [inline app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/OdysseyNotificationManager.java]: You should create the public part after you call setShowWhen. Currently your "private notification" has now a timestamp.

**Comment by djselbeck** [comment]
I'll test it, when it compiles again ;)

File: app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/OdysseyNotificationManager.java (hunk 3 above)
Comment (dennisguse) [reply app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/OdysseyNotificationManager.java]: I removed the Timestamp in commit: https://github.com/gateship-one/odyssey/commit/66075a411864e045f4669d25d196015fc8e42b20

This change was already merged into master.

I thought that a time is not really useful on a music player notification, as the notification will

**Comment by dennisguse** [comment]
Works again.
And removed the modifier change of gradlew.
It was for my convenience only :)

**Comment by djselbeck** [comment]
A few things:

- I think the "Content hidden" text is superfluous. I checked with several other applications: FB Messenger Lite, AOSP SMS, Telegram, K9-mail all show just nothing.

- The setShowWhen call needs to be called BEFORE the first notification is build. Otherwise it is useless.

**Comment by dennisguse** [comment]
**SetShowWhen** got it -  fixed it.

**About "Contents hidden"** 
I actually copied the behavior from K9Mail (version 5.403) and the default messaging app (com.android.messaging // version 1.0.001), which I use on LineageOS 7.1.2.
I attached some screenshots. 

If we don't show "Contents hidden", some user might mistake this as a bug.

screenshot_20180221-001525
screenshot_20180221-001402

**Comment by dennisguse** [comment]
"Contents hidden" is set automatically by Android if _no_ public notification is set.
However, then also the playbacks controls are not shown.

Next step is to find, how to apply this hidden-style for custom notifications...

**Comment by dennisguse** [comment]
So, it's done.
A friend (@frahaase) helped me to discover how to import Android system-defined strings.
The implemented code now uses the Android's default string - it should be thus consistent on each device.

Tested on LineageOS 7.1 only.

**Comment by frahaase** [comment]
You can find the documentation here:
https://developer.android.com/reference/android/content/res/Resources.html
Get systems returns:
Return a global shared Resources object that provides access to only system resources (no application resources), and is not configured for the current screen (can not use dimension units, does not change based on orientation, etc).

**Comment by djselbeck** [comment]
I've also found this but this provides no documentation for the strings used for this method. 

https://developer.android.com/reference/android/content/res/Resources.html#getIdentifier(java.lang.String,%20java.lang.String,%20java.lang.String) 

I mean is there some documentation for the use of package "android"?

**Comment by dennisguse** [comment]
Sadly, there is not documentation about this anywhere to find.
We only found that the strings are actually part of the AndroidSDK.

@djselbeck 
What do you think about using this approach, but check if the resource is available (check getIdentifer() != 0) and if it is not available, just use an empty string?

**Comment by gnome17** [comment]
We don't think it is a good solution to simply use some strings that are part of the SDK but not directly accessible. So please switch back to your previous solution with an additional string in Odyssey's resources.

**Comment by dennisguse** [comment]
I reverted the last change and now Odysseys strings.xml is used again.

**Comment by djselbeck** [comment]
I'm currently somewhat disconnected from the internet. I'll check and merge it as soon as I got a stable connection again

**Comment by dennisguse** [comment]
@djselbeck Any news?

**Comment by gnome17** [comment]
Hi,
sorry for the delay we have currently not much time but I will try to get back to this pr in the next weeks.

**Comment by gnome17** [comment]
Hi,
I finally tested your pr, works well on Android 6 and 8 and code looks fine.
But I noticed that in the lockscreen you will still have the cover of the current album as the background. Was this intended? It seems a bit weird to hide the notification if you can gather most of the information from the cover anyway.

**Comment by dennisguse** [comment]
You are definitely right: it absolutely makes sense to also hide the artwork on the lock screen.
I overlooked this completely.

Do you know if the lock screen art can be hidden while it is shown everywhere else?
If the metaDataBuilder is used only for the lock screen, then it might be sufficient to adjust it in PlaybackServiceHelper, or?

https://github.com/gateship-one/odyssey/blob/dee3531c3c2fe35fab49408ac7ac0164cee2e1b6/app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java#L370

https://github.com/gateship-one/odyssey/blob/dee3531c3c2fe35fab49408ac7ac0164cee2e1b6/app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java#L222

**Comment by gnome17** [comment]
I think it should be sufficient to adjust the two lines in the PlaybackServiceHelper that you mentioned.

**Comment by dennisguse** [comment]
It is working, but I didnt have the time to refactor the patch, so the names of variables/methods etc. matches the updated functionality.

File: app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java [hunk 4]
```diff
@@ -400,6 +407,15 @@ public void hideArtwork(boolean enable) {
         updateStatus();
     }
 
+    public void notificationPrivate(boolean enable) {
+      mNotificationPrivate = enable;
+      mNotificationManager.notificationPrivate(enable);
+
+//      this.hideArtwork(mHideArtwork);
```
Comment (gnome17) [inline app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java]: please remove this

File: app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java [hunk 5]
```diff
@@ -400,6 +407,15 @@ public void hideArtwork(boolean enable) {
         updateStatus();
     }
 
+    public void notificationPrivate(boolean enable) {
```
Comment (gnome17) [inline app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java]: please add a comment for this method like it is done for the other methods

File: app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java [hunk 6]
```diff
@@ -367,7 +371,10 @@ public void receiveAlbumBitmap(Bitmap bm) {
             // Try to get old metadata to save image retrieval.
             MediaMetadataCompat.Builder metaDataBuilder;
             metaDataBuilder = new MediaMetadataCompat.Builder(mMediaSession.getController().getMetadata());
-            metaDataBuilder.putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, bm);
+
+            if (mNotificationPrivate) metaDataBuilder.putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, null);
```
Comment (gnome17) [inline app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/PlaybackServiceStatusHelper.java]: even if it is one single line please use brackets

File: app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/OdysseyNotificationManager.java [hunk 7]
```diff
@@ -241,6 +254,17 @@ public void hideArtwork(boolean enable) {
         mHideArtwork = enable;
     }
 
+    public void notificationPrivate(boolean enable) {
```
Comment (gnome17) [inline app/src/main/java/org/gateshipone/odyssey/playbackservice/managers/OdysseyNotificationManager.java]: please add a comment for this method like it is done for the other methods

File: app/src/main/java/org/gateshipone/odyssey/playbackservice/PlaybackService.java [hunk 8]
```diff
@@ -1667,6 +1667,10 @@ public void hideArtwork(boolean enable) {
         mPlaybackServiceStatusHelper.hideArtwork(enable);
     }
 
+    public void notificationPrivate(boolean enable) {
```
Comment (gnome17) [inline app/src/main/java/org/gateshipone/odyssey/playbackservice/PlaybackService.java]: please add a comment for this method like it is done for the other methods

**Comment by dennisguse** [comment]
I changed the pref, methods and variables to _hideMediaOnLockscreen_ (more or less).
What do you think?

**Comment by gnome17** [comment]
Hi,
looks good now, thanks for your contribution.
If you want we will add you to the list of contributors. In this case we need to know if it is fine and if we should use your nickname or fullname provided by your github account or something different.

**Comment by dennisguse** [comment]
Great and thanks for the offer.
Feel free to decide it on your own - it was just one patch and nothing too big.
If you decide to then you can use my full name, which is also my Github username :)

PS Thanks for your feedback to the patch and for creating Odysseus in the first place.

**Pull request merged**